README
NAME

    Text::Levenshtein::Damerau::XS - Calculate edit distance based on
    insertion, deletion, substitution, and transposition

VERSION

    version 4.00

SYNOPSIS

        use Text::Levenshtein::Damerau::XS qw/lddistance/;
    
        print lddistance('Neil','Niel');
        # prints 1

DESCRIPTION

    Returns the true Damerau Levenshtein edit distance of strings with
    adjacent transpositions. XS implementation (requires a C compiler).
    Works correctly with utf8.

        use Text::Levenshtein::Damerau::XS qw/lddistance/;
        use utf8;
    
        lddistance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ'), 
        # prints 1

    Speed improvements over Text::Levenshtein::Damerau::PP:

        # Text::Levenshtein::Damerau::PP::pp_edistance("four","fuor")
        timethis 1000000: 381 wallclock secs (380.45 usr +  0.01 sys = 
             380.46 CPU) @ 2628.40/s (n=1000000)
    
        # Text::Levenshtein::Damerau::XS::lddistance("four","fuor")
        timethis 1000000: 19 wallclock secs (19.43 usr +  0.00 sys = 
             19.43 CPU) @ 51466.80/s (n=1000000)

METHODS

 lddistance

    Arguments: $source_text, $target_text, (optional) $max_distance

    Return value: Int $edit_distance || undef (if max_distance is exceeded)

    Returns: int that represents the edit distance between the two
    argument, or undef if $max_distance threshold is exceeded.

    Wrapper function to take the edit distance between a source and target
    string using XS algorithm implementation.

        use Text::Levenshtein::Damerau::XS qw/lddistance/;
        print lddistance('Neil','Niel');
        # prints 1
    
        # Set a max edit distance of 1
        print lddistance('Neil','Niely',1); # distance is 2
        # prints -1

TODO

      * Handle very large strings of text. Can be accomplished by reworking
      the scoring matrix or writing to disk.

      * Add from_file methods.

      * Add binary/byte string support.

SEE ALSO

      * Text::Levenshtein::Damerau

      * Text::Levenshtein::Damerau::PP

REPOSITORY

    https://github.com/ugexe/Text--Levenshtein--Damerau--XS

BUGS

    Please report bugs to:

    https://github.com/ugexe/Text--Levenshtein--Damerau--XS/issues

AUTHOR

    ugexe <ugexe@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Nick Logan.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision history for Perl module Text::Levenshtein::Damerau::XS

4.00 2014-10-20 21:47:54 America/New_York
    - Major changes: use lddistance instead of xs_edistance and return undef instead of -1 for > $max_distance
    - Use Dist::Zilla

3.0 2013-06-22 22:29:17 America/New_York
    - maxDistance incorrect results patched with non-optimal solution
    - some XS API changes to try and pass weird OpenBSD failures
    
2.9 2013-06-22 14:39:31 2013 America/New_York
    - Hacked in fix for maxDistance, needs to be redone optimally 

2.8 2013-01-25 10:41:04 2013 America/New_York
    - Removed unused function causing problems on certain windows builds.

2.7 2013-01-23 10:52:38 2013 America/New_York
    - Fixed memory leak from alloca->malloc switch

2.6 2013-01-19 12:09:57 2013 America/New_York
    - Changed alloca to malloc for source/target strings

2.5 2013-01-17 16:19:20 2013 America/New_York
    - Fixed accidental removal of MAX macro last release

2.4 2013-01-16 14:52:44 2013 America/New_York
    - use XSLoader when possible
    - slight reordering of operations

2.3 2013-01-13 20:00:09 2013 America/New_York
    - bulk88 optimizations
    - Removed lexicals from XS.pm
    - csx_edistance no longer exportable
    - C89; compiles under MSVC

2.2 2013-01-10 23:52:34 2013 America/New_York
    - NO_XSLOCKS to let malloc work correctly

2.1 2013-01-08 20:18:02 2013 America/New_York
    - Fixed pod error
    - Updated docs
    - examples/fuzzy_name_search.pl

2.0 2013-01-08 16:54:40 2013 America/New_York
    - Refactoring/optimizations
    - Better memory management
    - TODO added to docs

1.91 2013-01-03 16:16:21 2013 America/New_York
    - bulk88 C optimizations
    - variable/function name changes

1.9 2013-01-02 11:53:58 2013 America/New_York
    - Memory management refactoring
    - changes from Mithaldu/bulk88 merged

1.8 2013-01-13 19:07:47 America/New_York
    - Fixed bug created by last revision string check

1.7 2013-01-13 00:39:41 America/New_York
    - Matching string check before calculations take place
    - MIN_PERL_VERSION added back to Makefile using different number format

1.6 2013-01-06 15:52:57 America/New_York
    - Minmum Perl version supported changed from 5.8.0 to 5.8.8 America/New_York
1.5 2012-12-03 16:01:11 America/New_York
    - Removed MIN_PERL_VERSION from makefile, fixing netbsd error

1.4 2012-11-21 21:01:05 America/New_York
    - XS more XSish and less Cish
    - Memory usage improved

1.3 2012-11-21 12:49:32 2012 America/New_York
    - Removed built objects preventing install 

1.2 2012-11-20 00:35:20 2012 America/New_York
    - Added minimum Perl requirement of 5.8.0 due to passing AV*s not being allowed in prior versions. 

1.1 2012-11-18 18:47:43 2012 America/New_York
    - Moved XS.pm to lib/Text/Levenshtein/Damerau/XS.pm

1.0 2012-11-18 18:15:17 2012 America/New_York
    - Moved pod tests to /xt
    - Doc updates

0.9 2012-11-13 17:34:19 2012 America/New_York
    - max distance added to stop calculations at a certain distance
    - Code rehauled, thanks homercles
    - count[30000] (dictionary) replaced with linked list

0.8 2011-10-30 22:13:38 2012 America/New_York
    - Optimized memory usage
    - License set to perl in metas instead of LICENSE file only

0.7 2011-10-30 20:11:42 2012 America/New_York
    - Fixed pointer/int comparison warning
    - Documentation updates

0.6 2011-10-27 23:31:13 2012 America/New_York
    - Document updates

0.5 2012-10-27 18:29:57 2012 America/New_York
    - Fixed pod utf8 stuff

0.4 2012-10-27 15:58:22 2012 America/New_York
    - Document updates

0.3 2012-10-27 02:56:05 2012 America/New_York
    - Fixex examples in documentation.

0.2 2012-10-26 17:39:51 2012 America/New_York
    - Fixed problem installing from cpan.






lib/Text/Levenshtein/Damerau/XS.pm
package Text::Levenshtein::Damerau::XS;

# generated by Dist::Zilla::Plugin::OurPkgVersion
# VERSION
# AUTHORITY

use 5.008;
use strict;
use warnings FATAL => 'all';
require Exporter;
require XSLoader;
XSLoader::load('Text::Levenshtein::Damerau::XS', $Text::Levenshtein::Damerau::XS::VERSION);

@Text::Levenshtein::Damerau::XS::EXPORT_OK  = qw/xs_edistance lddistance/;
@Text::Levenshtein::Damerau::XS::ISA        = qw/Exporter/;



sub xs_edistance {
    my $distance = lddistance(@_);
    warnings::warnif("deprecated","xs_edistance is depreciated. Please use lddistance which returns undef instead of -1 for distances that exceed the optional \$max_distance");
    return (defined $distance && $distance >= 0) ? $distance : -1;
}



sub lddistance {
    # shift shift shift is faster than $_[0] $_[1] $_[2]
    return Text::Levenshtein::Damerau::XS::cxs_edistance( [unpack('U*', shift)], [unpack('U*',shift)], shift || 0);
}



1;



__END__




# ABSTRACT: Calculate edit distance based on insertion, deletion, substitution, and transposition

=for Pod::Coverage dl_load_flags cxs_edistance xs_edistance

=encoding utf8

=head1 SYNOPSIS

    use Text::Levenshtein::Damerau::XS qw/lddistance/;

    print lddistance('Neil','Niel');
    # prints 1

=head1 DESCRIPTION

Returns the true Damerau Levenshtein edit distance of strings with adjacent transpositions. XS implementation (requires a C compiler). Works correctly with utf8.

    use Text::Levenshtein::Damerau::XS qw/lddistance/;
    use utf8;

    print lddistance('ⓕⓞⓤⓡ','ⓕⓤⓞⓡ');
    # prints 1

Speed improvements over L<Text::Levenshtein::Damerau::PP>:

    # Text::Levenshtein::Damerau::PP::pp_edistance("four","fuor")
    timethis 1000000: 381 wallclock secs (380.45 usr +  0.01 sys =
         380.46 CPU) @ 2628.40/s (n=1000000)

    # Text::Levenshtein::Damerau::XS::lddistance("four","fuor")
    timethis 1000000: 19 wallclock secs (19.43 usr +  0.00 sys =
         19.43 CPU) @ 51466.80/s (n=1000000)

=head1 METHODS

=head2 lddistance

=over 4

=item Arguments: $source_text, $target_text, (optional) $max_distance

=item Return value: Int $edit_distance || undef (if max_distance is exceeded)

=back

Returns: int that represents the edit distance between the two argument, or undef if $max_distance threshold is exceeded.

Wrapper function to take the edit distance between a source and target string using XS algorithm implementation.

    use Text::Levenshtein::Damerau::XS qw/lddistance/;
    print lddistance('Neil','Niel');
    # prints 1

    # Set a max edit distance of 1
    print lddistance('Neil','Niely',1); # distance is 2
    # prints -1

=head1 TODO

=over 4

=item * Handle very large strings of text. Can be accomplished by reworking the scoring matrix or writing to disk.

=item * Add from_file methods.

=item * Add binary/byte string support.

=back

=head1 SEE ALSO

=over 4

=item * L<Text::Levenshtein::Damerau>

=item * L<Text::Levenshtein::Damerau::PP>

=back

=head1 REPOSITORY

L<https://github.com/ugexe/Text--Levenshtein--Damerau--XS>

=head1 BUGS

Please report bugs to:

L<https://github.com/ugexe/Text--Levenshtein--Damerau--XS/issues>

=cut
